### PR TITLE
add relay4 support

### PR DESCRIPTION
**Pull request by thetoster**
Signed-off-by: bartlomiej.zarnowski <bartlomiej.zarnowski@imgtec.com>

### CODE REVIEW

File: include/letmecreate/click/relay4.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+ * @return 0 if success otherwise -1
+ */
+uint8_t relay4_enable(uint8_t relay0Pin, uint8_t relay1Pin, uint8_t relay2Pin, uint8_t relay3Pin);
+
+/**
+ * @brief Same as calling relay4_enable(GPIO_83, GPIO_81, GPIO_80, GPIO_72)
+ *
+ * @return 0 if success otherwise -1
+ */
+uint8_t relay4_enable_defaults();
```
Comment (mtusnio) [inline include/letmecreate/click/relay4.h]: I'd get rid of this function entirely and maybe just document the preferred pins in relay4_enable

File: include/letmecreate/click/relay4.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+ INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+ SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE 
+ USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+************************************************************************************************************************/
+
+#ifndef RELAY4_H_
+#define RELAY4_H_
```
Comment (mtusnio) [inline include/letmecreate/click/relay4.h]: Just as with CO click, the header guard should be along the lines of __LETMECREATE_CLICK_RELAY_H__

**Comment by mtusnio** [comment]
Just one more thing, the relay4 is not strictly a click. Do we want a separate category for these ("boards"? "misc"?) @francois-berder ?

**Comment by francois-berder** [comment]
Indeed it is not a click board so it should be in a different part of the library. What about mikroe-others ?

File: src/click/relay4.c [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            }
+            g_Relay4_pins[t] = 0;
+        }
+    }
+    return result;
+}
+
+uint8_t relay4_set_state(uint8_t pinIndex, bool on) {
+    if ((pinIndex > 3) || (g_Relay4_pins[pinIndex] == 0)) {
+        fprintf(stderr, "Wrong pin Index, or pin is not initialized through call to relay4_enable.\n");
```
Comment (francois-berder) [inline src/click/relay4.c]: For each of your error message, prefix with "relay4: ".

File: src/click/relay4.c [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        return -1;
+
+    } else {
```
Comment (francois-berder) [inline src/click/relay4.c]: no need for else

File: src/click/relay4.c [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        return gpio_set_value(g_Relay4_pins[pinIndex], on);
+    }
+}
+
+uint8_t relay4_get_state(uint8_t pinIndex, bool* state) {
+    if ((pinIndex > 3) || (g_Relay4_pins[pinIndex] == 0)) {
+        fprintf(stderr, "Wrong pin Index, or pin is not initialized through call to relay4_enable.\n");
+        return -1;
+
+    } else {
```
Comment (francois-berder) [inline src/click/relay4.c]: no need for else

File: src/click/relay4.c [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        uint8_t intState;
+        uint8_t result = gpio_get_value(g_Relay4_pins[pinIndex], &intState);
+        if (state != NULL) {
```
Comment (francois-berder) [inline src/click/relay4.c]: move this check just after checking pinIndex

File: src/click/relay4.c [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        return result;
+    }
+}
+
+uint8_t relay4_toggle(uint8_t pinIndex, bool* newState) {
+    if ((pinIndex > 3) || (g_Relay4_pins[pinIndex] == 0)) {
+        fprintf(stderr, "Wrong pin Index, or pin is not initialized through call to relay4_enable.\n");
+        return -1;
+
+    } else {
```
Comment (francois-berder) [inline src/click/relay4.c]: no need for else

File: src/click/relay4.c [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        uint8_t intState;
+        uint8_t result = gpio_get_value(g_Relay4_pins[pinIndex], &intState);
+        if (result < 0) {
```
Comment (francois-berder) [inline src/click/relay4.c]:
```c
if (gpio_get_value(...) < 0)
    return -1
```

File: src/click/relay4.c [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+            return -1;
+        }
+        //note: new state is negation of current state
+        bool state = intState == 0 ? true : false;
+        if (newState != NULL) {
```
Comment (francois-berder) [inline src/click/relay4.c]: move it just after checking pinIndex

**Comment by mtusnio** [comment]
Alternatively any other boards/hardware can be added with just the platform prefix. Then we'll have a structure of:

letmecreate/core/
letmecreate/click/
letmecreate/mikro-e/
letmecreate/grove/

etc.

Clicks can be separated from Mikro-e since they are a special case for Creator.

**Comment by francois-berder** [comment]
Sounds good Michał.

**Comment by thetoster** [comment]
I disagree with move of 
'if (newState != NULL) {'

condition into different place, it breaks logic flow.

**Comment by mtusnio** [comment]
Can you move the relay4 implementation to letmecreate/mikro-e as per the discussion above?

**Comment by francois-berder** [comment]
@thetoster In the rest of the code, I first check the parameters, then the logic of the function follows. That's why I would prefer if you change your code. Same for the 
```c
if (gpio_get_value(...) < 0)
    return -1;
```

**Pull request merged**